gramlib/grammar.ml

```diff
@@ -871,7 +871,7 @@ and print_rule : type s tr p. formatter -> (s, tr, p) ty_symbols -> unit =
 and print_level : type s. _ -> _ -> s ex_symbols list -> _ =
   fun ppf pp_print_space rules ->
   fprintf ppf "@[<hov 0>[ ";
-  let _ =
+  let _ : _ -> _ =
```

Comment:
What's going on here?

Reply:
Last time I got it wrong, the fold threads a separator which is of type `formatter -> unit`, so it prints, then it discards the last value. This is done as to provide an empty separator for the first element of the list.

Reply:
Shouldn't we use `List.iteri` and test `if i = 0` instead then?

Reply:
That's another choice I guess, that code comes from camlp5.

Reply:
cf https://github.com/coq/coq/pull/10482